pilot/pkg/networking/v1alpha3/listener.go

```diff
@@ -603,3 +620,137 @@ func buildInboundHTTPRouteConfig(instance *model.ServiceInstance) *xdsapi.RouteC
+		if port == 0 {
+			nameToServiceMap[svc.Hostname] = svc
+		} else {
+			if svcPort, exists := svc.Ports.GetByPort(port); exists {
+				nameToServiceMap[svc.Hostname] = &model.Service{
+					Hostname: svc.Hostname,
+					Address:  svc.Address,
+					Ports:    []*model.Port{svcPort},
+				}
+			}
+		}
+	}
+
+	guardedHosts := TranslateVirtualHosts(env.VirtualServices(), nameToServiceMap, nil, node.Domain)
+	vHostPortMap := make(map[int][]route.VirtualHost)
+
+	// there should be only one guarded host in the return val since we supplied services with just one port
+	for _, guardedHost := range guardedHosts {
+		routes := make([]route.Route, 0)
+		for _, r := range guardedHost.Routes {
+			routes = append(routes, r.Route)
+		}
+
+		virtualHosts := make([]route.VirtualHost, 0)
+
+		for _, host := range guardedHost.Hosts {
+			virtualHosts = append(virtualHosts, route.VirtualHost{
+				Name:    fmt.Sprintf("%s:%d", host, guardedHost.Port),
+				Domains: []string{host},
+				Routes:  routes,
+			})
+		}
+
+		for _, svc := range guardedHost.Services {
+			domains := generateAltVirtualHosts(svc, guardedHost.Port)
+			if len(svc.Address) > 0 {
+				// add a vhost match for the IP (if its non CIDR)
+				cidr := convertAddressToCidr(svc.Address)
+				if cidr.PrefixLen.Value == 32 {
+					domains = append(domains, svc.Address)
+					domains = append(domains, fmt.Sprintf("%s:%d", svc.Address, guardedHost.Port))
+				}
+			}
+			virtualHosts = append(virtualHosts, route.VirtualHost{
+				Name:    fmt.Sprintf("%s:%d", svc.Hostname, guardedHost.Port),
+				Domains: domains,
+				Routes:  routes,
+			})
+		}
+
+		vHostPortMap[guardedHost.Port] = virtualHosts
+	}
+
+	var virtualHosts []route.VirtualHost
+	if routeName == RDSHttpProxy {
+		virtualHosts = mergeAllVirtualHosts(vHostPortMap)
+	} else {
+		virtualHosts = vHostPortMap[port]
+	}
+
+	out := &xdsapi.RouteConfiguration{
+		Name:         fmt.Sprintf("%d", port),
+		VirtualHosts: virtualHosts,
+		ValidateClusters: &google_protobuf.BoolValue{
+			Value: false, // until we have rds
+		},
+	}
+	return out
+
+}
+
+// Given a service, and a port, this function generates all possible HTTP Host headers.
+// For example, a service of the form foo.local.campus.net on port 80 could be accessed as
+// http://foo:80 within the .local network, as http://foo.local:80 (by other clients in the campus.net domain),
+// as http://foo.local.campus:80, etc.
+// The logic of this function is as follows:
+// given a DNS name, it reverses it, and splits the string in K different ways (where K= 0 to num dots in name)
+// the last element of the K-way string split is the hostname that we need.
+// for example, reviews.default.svc.local yields local.svc.default.reviews (omitting char reversal)
+// which when split in 2 ways yields local, svc.default.reviews
+// when split in 3 ways yields, local.svc, default.reviews, and so on.
+// Take the last element and reverse the string again to get the correct host (e.g, reviews.default.svc)
```

Comment:
I don't think the string reversal is probably necessary. You should be able to split the host on `.` and then just index to get the last label.

Reply:
example with no string reversal:
```golang
func generateAltVirtualHosts(hostname string, port int) []string {
	vhosts := []string{hostname, fmt.Sprintf("%s:%d", hostname, port)}
	for i := len(hostname)-1; i>=0; i-- {
		if hostname[i] == '.' {
			variant := hostname[:i]
			variantWithPort := fmt.Sprintf("%s:%d", variant, port)
			vhosts = append(vhosts, variant)
			vhosts = append(vhosts, variantWithPort)
		}
	}
	return vhosts
}
```